src/Engine/Graphics/BspRenderer.cpp

```diff
@@ -6,31 +6,46 @@
 
 #include "Library/Logger/Logger.h"
 
+// TODO(yoctozepto): we should not see it here
 BspRenderer *pBspRenderer = new BspRenderer();
 
-//----- (004B0EA8) --------------------------------------------------------
-void BspRenderer::AddFaceToRenderList_d3d(int node_id, int uFaceID) {
-    int pTransitionSector;  // ax@11
-    // int dotdist;                              // edx@15
 
+//----- (004B0EA8) --------------------------------------------------------
+void BspRenderer::AddFace(int node_id, int uFaceID) {
+    // TODO(yoctozepto): does the below happen?
     if (uFaceID >= pIndoor->pFaces.size())
-        return;
+        return;  // face out of range
+
     BLVFace *pFace = &pIndoor->pFaces[uFaceID];
 
     if (!pFace->isPortal()) {
+        // NOTE(yoctozepto): the below happens, e.g., on various stairs
+        if (pFace->Invisible())
+            return;  // nothing to render
+        // NOTE(yoctozepto): the below happens, e.g., on various stairs
+        if (pFace->uNumVertices < 3) {
+            return;  // nothing to render
+        }
+        // TODO(yoctozepto): does the below happen?
+        if (!pFace->GetTexture()) {
+            return;  // nothing to render
+        }
         if (num_faces < 1500) {
             // add face and return
             faces[num_faces].uFaceID = uFaceID;
             faces[num_faces++].uNodeID = node_id;
         } else {
+            // TODO(yoctozepto): does the below happen?
```

Comment:
Maybe could see it with badly set up portals. Definitely will see it if you break the bsp renderer

Reply:
Yeah, that would do it quickly. OK: if you don't recall this one either, I will make it an implicit assert from the array.